### PR TITLE
1.0.0

### PR DESCRIPTION
**Pull request by splatteredbits**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
+
+# 1.0.0
+
+The `Carbon.Windows.Installer` module was created from functions in the `Carbon` module. These release notes assume
+you're migrating from `Carbon`. If you're not migrating, you can ignore these release notes.
```
Comment (decoyjoe) [inline CHANGELOG.md]: Nit: 'Carbon' should be a link.

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  `Carbon.Windows.Installer.Records.Property` objects (instead of a hashtable) and we added a `GetPropertyValue` method
+  to lookup specific property values.
+* Remove usages of the `[Carbon.Msi.MsiInfo]` type. It was removed. `Get-CMsi` now returns `[PSObject]` objects with a
+  pstypename of `[Carbon.Windows.Installer.MsiInfo]`.
+
+### Get-CProgramInstallInfo
+
+* Rename usages of `Get-CProgramInstallInfo` to `Get-CInstalledProgram`. We renamed `Get-CProgramInstallInfo` to 
+  `Get-CInstalledProgram`. The `C` prefix is now required.
+* Remove usages of the `[Carbon.Computer.ProgramInstallInfo]`. `Get-CInstalledProgram` now returns `[PSObject]` objects
```
Comment (decoyjoe) [inline CHANGELOG.md]: Typo: add the word "type" after [Carbon.Computer.ProgramInstallInfo]

File: CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+* `Install-CMsi` can now download an MSI file to install. Pass the URL to the installer to the `Url` parameter, the
+file's SHA256 checksum to the `Checksum` parameter, its product name to the `ProductName` parameter, and its 
+product code to the `ProductCode` parameter. The MSI file will only be downloaded if it isn't installed. You can get the MSI's checksum with PowerShell's `Get-FileHash` cmdlet. You can get the MSI's product name and code with this module's `Get-CMsi` function.
+* Added `ArgumentList` parameter to `Install-CMsi` to pass additional arguments to `msiexec.exe` when installing an MSI.
+* Added `DisplayOption`, `LogOption`, and `LogPath` parameters to allow users to control the display and logging options passed to `msiexec.exe`.
+
+### Changed
+
+* Minimum system requirements are now Windows PowerShell 5.1 on .NET 4.5.2+ or PowerShell 6.2+ on Windows. The
+`Get-CMsi` and `Install-CMsi` functions don't work in PowerShll 6.2.
```
Comment (decoyjoe) [inline CHANGELOG.md]: Typo: `PowerShll` is missing an `e`

File: Carbon.Windows.Installer/Carbon.Windows.Installer.psd1 [hunk 4]
```diff
@@ -36,7 +36,28 @@
...
+graphical user interface as objects and text.
+
+Functions include:
+
+* `Get-CInstalledProgram`: reads the registry to determine all the programs installed on the local computer. Returns
+an object for each program. Is an object-based and text-based version of Windows' "Programs and Features"/"Apps and
+Features" GUI.
+* `Get-CMsi`: reads an MSI file and returns an object exposing the MSI's internal tables, like product name, product
+code, product version, etc. Let's you inspect an MSI for its metadata without installing it. Can also download the MSI
+file first. This function requires Windows PowerShell or PowerShell 7.1+ on Windows.
```
Comment (decoyjoe) [inline Carbon.Windows.Installer/Carbon.Windows.Installer.psd1:51]: Should `This function requires Windows PowerShell or PowerShell 7.1+ on Windows.` say `Powershell 6.2+`, per the minimum system requirements?

File: Carbon.Windows.Installer/Functions/Get-InstalledProgram.ps1 [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    * The `User` property will only be set for software installed for specific users. For global software, the `User`
+      property will be `[String]::Empty`.
+    * The `InstallDate` property is set to `[DateTime]::MinValue` if the install date can't be determined.
+    * The `Version` property is `$null` if the version can't be parsed.
+    
+    .EXAMPLE
+    Get-CInstalledProgram | Sort-Object 'DisplayName'
+
+    Demonstrates how to get a list of all the installed programs, similar to what the Programs and Features settings UI
+    shows. The returned objects are not sorted, so you'll usually want to pipe the output to `Sort-Object`.
```
Comment (decoyjoe) [inline Carbon.Windows.Installer/Functions/Get-InstalledProgram.ps1:38]: Seems like the default should be to sort by DisplayName like the "Programs and Features" applet.

File: Carbon.Windows.Installer/Functions/Get-InstalledProgram.ps1 [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+            [Version]$version = $null
+            if( [Version]::TryParse($rawVersion, [ref]$version) )
+            {
+                $info.Version = $version
+            }
+
+            $windowsInstallerValue = Get-KeyIntValue -Key $key -ValueName 'WindowsInstaller'
+            $info.WindowsInstaller = ($windowsInstallerValue -gt 0)
+
+            $info.pstypenames.Insert(0, 'Carbon.Windows.Installer.ProgramInfo')
```
Comment (decoyjoe) [inline Carbon.Windows.Installer/Functions/Get-InstalledProgram.ps1:237]: What a great idea, never thought about doing this. 👍

File: Carbon.Windows.Installer/Functions/Get-Msi.ps1 [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    Gets information from an MSI.
+
+    .DESCRIPTION
+    The `Get-CMsi` function uses the `WindowsInstaller.Installer` COM API to read properties from an MSI file. Pass the
+    path to the MSI file or files to the `Path`, or pipe file objects to `Get-CMsi`. An object is returned that exposes
+    the internal metadata of the MSI file:
+    
+    * `ProductName`, the value of the MSI's `ProductName` property.
+    * `ProductCode`, the value of the MSI's `ProductCode` property as a `Guid`.
+    * `ProduceLanguage`, the value of the MSI's `ProduceLanguage` property, as an integer.
```
Comment (decoyjoe) [inline Carbon.Windows.Installer/Functions/Get-Msi.ps1]: Typo: `ProduceLanguage` (🥬) should be `ProductLanguage`

File: Carbon.Windows.Installer/Functions/Get-Msi.ps1 [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+    tables, so by default, `Get-CMsi` only returns the records from the `Property` and `Feature` tables. The `Property`
+    table contains program metadata like product name, product code, etc. The `Feature` table contains the feature names
+    of any optional features you might want to install. When installing, these feature names would get passed to the
+    `msiexec` install command as a comma-separated list as the `ADDLOCAL` property, e.g. ADDLOCAL="Feature1,Feature2".
+
+    To return the records from additional tables, pass the table name or names to the `IncludeTable` parameter.
+    Wildcards supported. Records from the `Property` and `Feature` tables are *always* returned. The `TableNames`
+    property on returned objects is the list of all tables in the MSI's database.
+
+    Because this function uses the Windows Installer COM API, it requires Windows PowerShell 5.1 or PowerShell 7.1+ on
```
Comment (decoyjoe) [inline Carbon.Windows.Installer/Functions/Get-Msi.ps1:33]: The module docs say the minimum is PowerShell 6.2 (not 7.1).

File: Carbon.Windows.Installer/Functions/Get-Msi.ps1 [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+                    {
+                        return
+                    }
+
+                    $this.Property | Where-Object 'Property' -eq $Name | Select-Object -ExpandProperty 'Value'
+                }
+
+            $installer = New-Object -ComObject 'WindowsInstaller.Installer'
+
+            $database = $null;
```
Comment (decoyjoe) [inline Carbon.Windows.Installer/Functions/Get-Msi.ps1]: Errant semicolon.

File: Carbon.Windows.Installer/Functions/Get-Msi.ps1 [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+                    [void][Runtime.InteropServices.Marshal]::ReleaseCOMObject($installer)
+                    $collect = $true
+                }
+
+                if( $collect )
+                {
+                    # ReleaseCOMObject still leaves the MSI file open. The only way to close the file handle is to run
+                    # garbage collection, and even then it takes a few seconds. :(
+                    Debug "[GC]::Collect()  START"
+                    [GC]::Collect()
```
Comment (decoyjoe) [inline Carbon.Windows.Installer/Functions/Get-Msi.ps1:272]: Hopefully this doesn't cause any unintended side effects in the rest of the code running in the current PowerShell session.

File: Carbon.Windows.Installer/Functions/Install-Msi.ps1 [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+        # Install the MSI even if it has already been installed. Will cause a repair/reinstall to run.
+        [Switch] $Force,
+
+        # Controls how the MSI UI is displayed to the user. The default is `Quiet`, meaning no UI is shown. Valid values
+        # are `Passive`, a UI showing a progress bar is shown, or `Full`, the UI is displayed to the user as if they
+        # double-clicked the MSI file.
+        [ValidateSet('Quiet', 'Passive', 'Full')]
+        [String] $DisplayMode = 'Quiet',
+
+        # The logging options to use. The default is to log all information (`*`), log verbose output (`v`), log exta
```
Comment (decoyjoe) [inline Carbon.Windows.Installer/Functions/Install-Msi.ps1]: Typo: `exta` should be `extra`

**Review by decoyjoe — CHANGES_REQUESTED** [review]

File: Carbon.Windows.Installer/Carbon.Windows.Installer.psd1 (hunk 4 above)
Comment (splatteredbits) [reply Carbon.Windows.Installer/Carbon.Windows.Installer.psd1:51]: PowerShell 6.X doesn't support COM on Windows and those functions use the COM installer API.

File: Carbon.Windows.Installer/Functions/Get-InstalledProgram.ps1 (hunk 5 above)
Comment (splatteredbits) [reply Carbon.Windows.Installer/Functions/Get-InstalledProgram.ps1:38]: This function already takes a little over 2 seconds:

```
$ 1..10 | % { measure-command { Get-CInstalledProgram } } | ft Minutes,Seconds,Milliseconds

Minutes Seconds Milliseconds
------- ------- ------------
      0       2          166
      0       2          184
      0       2          186
      0       2          249
      0       2          586
      0       2          239
      0       2          272
      0       2          199
      0       2          187
      0       2          170


```

If I add a sort, it looks like it takes an extra quarter to half a second:

```
Minutes Seconds Milliseconds
------- ------- ------------
      0       2          841
      0       2          462
      0       2          573
      0       2          822
      0       2          499
      0       2          335
      0       2          163
      0       2          170
      0       2          607
      0       2          529
```

Comment (decoyjoe) [reply Carbon.Windows.Installer/Functions/Get-InstalledProgram.ps1:38]: Yikes. Definitely don't want to pre-sort then.

Comment (splatteredbits) [reply Carbon.Windows.Installer/Functions/Get-InstalledProgram.ps1:38]: I spent a little time optimizing this function, and got the timings down almost 50%, including the sort (which adds about 1/4 second).

```
$ % { measure-command { Get-CInstalledProgram } } | ft Minutes,Seconds,Milliseconds

Minutes Seconds Milliseconds
------- ------- ------------
      0       1          906
      0       1          134
      0       1          191
      0       1          167
      0       1          170
      0       1          129
      0       1           94
      0       1          100
      0       1          150
      0       1          134


```

File: Carbon.Windows.Installer/Functions/Get-Msi.ps1 (hunk 10 above)
Comment (splatteredbits) [reply Carbon.Windows.Installer/Functions/Get-Msi.ps1:272]: It shouldn't. At worse, the function will take a little longer while the garbage collector runs.

**Review by decoyjoe — APPROVED** [review]

**Pull request merged**